main/src/library/Result.flix

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2016 Liam Palmer
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Result {
+
+    ///
+    /// Returns `true` iff `r` is `Ok(v)`.
+    ///
+    def isOk[a, b](r: Result[a, b]): Bool = match r with {
+        case Ok(v) => true
+        case Err(e) => false
+    }
+
+    ///
+    /// Returns `true` iff `r` is `Err(e)`.
+    ///
+    def isErr[a, b](r: Result[a, b]): Bool = match r with {
+        case Ok(v) => false
+        case Err(e) => true
+    }
+
+    ///
+    /// Returns `v` if `r` is `Ok(v)`. Undefined semantics if `r` is `Err(e)`.
+    ///
+    def ok[a, b](r: Result[a, b]): a = match r with {
+        case Ok(v) => v
+    }
+
+    ///
+    /// Returns `e` if `r` is `Err(e)`. Undefined semantics if `r` is `Ok(v)`.
+    ///
+    def err[a, b](r: Result[a, b]): b = match r with {
+        case Err(e) => e
+    }
+
+    ///
+    /// Returns `v` if `r` is `Ok(v)`. Otherwise returns `d`.
+    ///
+    def okWithDefault[a, b](r: Result[a, b], d: a): a = match r with {
+        case Ok(v) => v
+        case Err(e) => d
+    }
+
+    ///
+    /// Returns `e` if `r` is `Err(e)`. Otherwise returns `d`.
+    ///
+    def errWithDefault[a, b](r: Result[a, b], d: b): b = match r with {
```

Comment:
I am not sure "how many" of the regular operations we want for error. It would probably be useful to have a few, but lets decide on those last, i.e. you can postpone implementing them for now.